crates/forge/src/lib.rs

```diff
@@ -352,97 +361,144 @@ pub fn run(
-                package_root,
-                &runner,
-                case,
-                runner_config.fork_targets.as_ref(),
-                &runner_params.contracts,
-                &runner_params.predeployed_contracts,
-                vec![],
-                &runner_params.environment_variables,
-            )?;
-            pretty_printing::print_test_result(&result, None);
-
-            result
-        } else {
-            was_fuzzed = true;
-            let (result, runs) = run_with_fuzzing(
-                package_root,
-                runner_config,
-                runner_params,
-                &runner,
-                case,
-                &args,
-            )?;
-            pretty_printing::print_test_result(&result, Some(runs));
-
-            result
-        };
 
-        results.push(result.clone());
+        let function = runner.find_function(case_name).unwrap();
+        let args = function_args(function, &BUILTINS);
 
-        if runner_config.exit_first {
-            if let TestCaseSummary::Failed { .. } = result {
-                for case in &tests.test_cases[i + 1..] {
-                    let skipped_result = TestCaseSummary::skipped(case);
-                    pretty_printing::print_test_result(&skipped_result, None);
-                    results.push(skipped_result);
-                }
-                return Ok((
-                    TestCrateSummary {
-                        test_case_summaries: results,
-                        runner_exit_status: RunnerStatus::TestFailed,
-                        test_crate_type: tests.test_crate_type,
+        let task = task::spawn({
+            let package_root = package_root.clone();
+            let case = Arc::new(case.clone());
+            let c = case.clone();
+            let runner = runner.clone();
+            let args: Vec<ConcreteTypeId> = args.into_iter().cloned().collect();
+            let cancellation_token = cancellation_token.clone();
+            let runner_params = runner_params.clone();
+            let runner_config = runner_config.clone();
+            async move {
+                tokio::select! {
+                    _ = cancellation_token.cancelled() => {
+                        // The token was cancelled
+                       Ok((TestCaseSummary::skipped(&c), None))
+                    },
+                    result = run_p_test(package_root.clone() ,case, runner,runner_params, runner_config, args, cancellation_token.clone()) => {
+                        result
                     },
-                    was_fuzzed,
-                ));
+                }
             }
+        });
+
+        tasks.push(task);
+    });
+
+    let mut results = vec![];
+
+    for thread in tasks {
+        let (result, runs) = thread.await.unwrap()?;
+        if runs.is_some() {
+            was_fuzzed = true;
         }
+
+        pretty_printing::print_test_result(&result, runs);
+        results.push((result, runs));
     }
```

Comment:
Instead of doing this we should add `futures` dependency and

```rust
use futures::future::join_all;

join_all(tasks).await()
```

This loop is not an efficient way of waiting for multiple threads.

Reply:
I don't think it much matters performance wise, but is better to do it this way (`join_all` or `select_all`) as it is more idiomatic and easier to reason about.

Reply:
@MaksymilianDemitraszek isn't it blocking in ever loop run though? Unless tokio is smart and somehow compiles it to wait for all threads at once under the hood (like Django with for loop and db requests)

Reply:
Yea, it would block in single threaded async setting, but we use `task::spawn` which is immediately executed on a separate thread.

Reply:
I found `use futures::stream::{FuturesOrdered, FuturesUnordered};` and it seems has better performance. Mostly FuturesUnordered which we are able to use in fuzzing tests